pkg/rpcserver/rpcserver.go

```diff
@@ -278,22 +295,44 @@ func (serv *server) handleRunnerConn(runner *Runner, conn *flatrpc.Conn) error {
 		opts.Features = serv.cfg.Features
 	}
 
-	err := runner.Handshake(conn, opts)
+	info, err := runner.Handshake(conn, opts)
 	if err != nil {
 		log.Logf(1, "%v", err)
 		return err
 	}
 
+	eg, ctx := errgroup.WithContext(baseCtx)
+
+	serv.checkOnce.Do(func() {
+		// Run the machine check.
+		eg.Go(func() error {
+			if err := serv.runCheck(ctx, &info); err != nil {
+				return fmt.Errorf("%w: %w", errFatal, err)
+			}
+			return nil
+		})
+	})
+
 	if serv.triagedCorpus.Load() {
-		if err := runner.SendCorpusTriaged(); err != nil {
-			log.Logf(2, "%v", err)
-			return err
-		}
+		eg.Go(runner.SendCorpusTriaged)
 	}
 
-	return serv.connectionLoop(runner)
+	stopCtx, stop := context.WithCancel(ctx)
+	go func() {
+		<-stopCtx.Done()
+		runner.Stop()
+	}()
+	defer stop()
+
+	eg.Go(func() error {
+		return serv.connectionLoop(runner)
+	})
+	return eg.Wait()
```

Comment:
We tied serv.runCheck to one random runner. This means this handleRunnerConn won't return until the check is done, even if the VM has crashed. Can this cause problems? When we have few (1-2) VMs?

Reply:
Good point!
I wanted to avoid creating an `errgroup.Group` in `Serve()` and passing it down as an argument, but it seems like it's the only reliable way. Updated the code to untie `runCheck` from the runner handler.